File: disc_golf_app/lib/screens/onboarding/onboarding_screen.dart
import 'package:flutter/material.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../home_screen.dart';

/// Five-page first-launch onboarding walkthrough.
/// Shown once; completion is persisted in SharedPreferences.
class OnboardingScreen extends StatefulWidget {
  const OnboardingScreen({super.key});

  @override
  State<OnboardingScreen> createState() => _OnboardingScreenState();
}

class _OnboardingScreenState extends State<OnboardingScreen> {
  final _controller = PageController();
  int _page = 0;

  static const _pages = [
    _OnboardingPage(
      icon: Icons.sports,
      color: Color(0xFF0f3460),
      title: 'Welcome to\nDisc Flight School',
      body:
          'Your personal disc golf coach — analyze your form, track disc flight, '
          'and practice smarter with guided challenges.',
      isFirst: true,
    ),
    _OnboardingPage(
      icon: Icons.accessibility_new,
      color: Color(0xFF16213e),
      title: 'Form Coach',
      body:
          'Record a throw and let AI-powered pose detection score your technique. '
          'Compare against pro baselines phase by phase and get targeted cues to fix '
          'the angles that matter most.',
    ),
    _OnboardingPage(
      icon: Icons.track_changes,
      color: Color(0xFF0f3460),
      title: 'Flight Tracker',
      body:
          'Place disc markers on your video frame by frame to map the exact '
          'flight path. Zoom in for precise placement, draw a target line, '
          'and export the full trajectory data.',
    ),
    _OnboardingPage(
      icon: Icons.casino,
      color: Color(0xFF16213e),
      title: 'Disc Roulette',
      body:
          'Spin the wheel to get a random disc, distance, and difficulty '
          'combination. Add hindrances or power boosts for extra challenge — '
          'great for structured field work.',
    ),
    _OnboardingPage(
      icon: Icons.menu_book,
      color: Color(0xFF0f3460),
      title: 'Knowledge Base',
      body:
          'Browse tips, drills, and technique articles — or ask the AI assistant '
          'a specific question. Everything you need to understand the "why" behind '
          'the cues.',
      isLast: true,
    ),
  ];

  Future<void> _finish() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setBool('onboarding_complete', true);
    if (!mounted) return;
    Navigator.of(context).pushReplacement(
      MaterialPageRoute(builder: (_) => const HomeScreen()),
    );
  }

  void _next() {
    if (_page < _pages.length - 1) {
      _controller.nextPage(
        duration: const Duration(milliseconds: 300),
        curve: Curves.easeInOut,
      );
    } else {
      _finish();
    }
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final isLast = _page == _pages.length - 1;

    return Scaffold(
      body: Stack(
        children: [
          PageView.builder(
            controller: _controller,
            itemCount: _pages.length,
            onPageChanged: (i) => setState(() => _page = i),
            itemBuilder: (_, i) => _pages[i],
          ),

          // Bottom nav
          Positioned(
            left: 0,
            right: 0,
            bottom: 36,
            child: Padding(
              padding: const EdgeInsets.symmetric(horizontal: 24),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: [
                  // Skip
                  TextButton(
                    onPressed: isLast ? null : _finish,
                    child: Text(
                      isLast ? '' : 'Skip',
                      style: const TextStyle(color: Colors.white54),
                    ),
                  ),

                  // Dot indicators
                  Row(
                    children: List.generate(
                      _pages.length,
                      (i) => AnimatedContainer(
                        duration: const Duration(milliseconds: 250),
                        margin:
                            const EdgeInsets.symmetric(horizontal: 4),
                        width: i == _page ? 18 : 8,
                        height: 8,
                        decoration: BoxDecoration(
                          color: i == _page
                              ? Colors.lightBlueAccent
                              : Colors.white24,
                          borderRadius: BorderRadius.circular(4),
                        ),
                      ),
                    ),
                  ),

                  // Next / Get Started
                  ElevatedButton(
                    onPressed: _next,
                    style: ElevatedButton.styleFrom(
                      backgroundColor: Colors.lightBlueAccent,
                      foregroundColor: Colors.black87,
                      padding: const EdgeInsets.symmetric(
                          horizontal: 20, vertical: 10),
                      shape: RoundedRectangleBorder(
                          borderRadius: BorderRadius.circular(20)),
                    ),
                    child: Text(isLast ? 'Get Started' : 'Next'),
                  ),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }
}

class _OnboardingPage extends StatelessWidget {
  final IconData icon;
  final Color color;
  final String title;
  final String body;
  final bool isFirst;
  final bool isLast;

  const _OnboardingPage({
    required this.icon,
    required this.color,
    required this.title,
    required this.body,
    this.isFirst = false,
    this.isLast = false,
  });

  @override
  Widget build(BuildContext context) {
    return Container(
      color: color,
      padding: const EdgeInsets.fromLTRB(32, 80, 32, 100),
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Container(
            width: 120,
            height: 120,
            decoration: BoxDecoration(
              color: Colors.white10,
              shape: BoxShape.circle,
              border:
                  Border.all(color: Colors.lightBlueAccent, width: 2),
            ),
            child: Icon(icon,
                size: 64, color: Colors.lightBlueAccent),
          ),
          const SizedBox(height: 40),
          Text(
            title,
            textAlign: TextAlign.center,
            style: const TextStyle(
              fontSize: 28,
              fontWeight: FontWeight.bold,
              color: Colors.white,
              height: 1.25,
            ),
          ),
          const SizedBox(height: 20),
          Text(
            body,
            textAlign: TextAlign.center,
            style: const TextStyle(
              fontSize: 16,
              color: Colors.white70,
              height: 1.5,
            ),
          ),
        ],
      ),
    );
  }
}

File: disc_golf_android/app/src/main/kotlin/com/discflightschool/app/ui/screens/OnboardingScreen.kt
package com.discflightschool.app.ui.screens

import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.pager.HorizontalPager
import androidx.compose.foundation.pager.rememberPagerState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.MenuBook
import androidx.compose.material.icons.filled.AccessibilityNew
import androidx.compose.material.icons.filled.Casino
import androidx.compose.material.icons.filled.SportsBaseball
import androidx.compose.material.icons.filled.TrackChanges
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.discflightschool.app.LocalAppContainer
import com.discflightschool.app.ui.theme.AppColors
import kotlinx.coroutines.launch

private data class OnboardingPage(
    val icon: ImageVector,
    val background: Color,
    val title: String,
    val body: String,
)

private val onboardingPages = listOf(
    OnboardingPage(
        icon = Icons.Default.SportsBaseball,
        background = AppColors.TopBar,
        title = "Welcome to\nDisc Flight School",
        body = "Your personal disc golf coach — analyze your form, track disc flight, " +
            "and practice smarter with guided challenges.",
    ),
    OnboardingPage(
        icon = Icons.Default.AccessibilityNew,
        background = AppColors.Surface,
        title = "Form Coach",
        body = "Record a throw and let on-device pose detection score your technique. " +
            "Compare against pro baselines phase by phase and get targeted cues to fix " +
            "the angles that matter most.",
    ),
    OnboardingPage(
        icon = Icons.Default.TrackChanges,
        background = AppColors.TopBar,
        title = "Flight Tracker",
        body = "Let the detector find the disc, or place markers yourself frame by frame " +
            "to map the exact flight path. Zoom in for precise placement, draw a target " +
            "line, and export the trajectory.",
    ),
    OnboardingPage(
        icon = Icons.Default.Casino,
        background = AppColors.Surface,
        title = "Disc Roulette",
        body = "Spin for a random disc, shot shape, and hindrance. Play a full scored " +
            "round where every stroke is weighted by how hard the challenge was.",
    ),
    OnboardingPage(
        icon = Icons.AutoMirrored.Filled.MenuBook,
        background = AppColors.TopBar,
        title = "Knowledge Base",
        body = "Browse tips, drills, and technique articles — or ask a question and get " +
            "an answer grounded in the peer-reviewed research bundled with the app.",
    ),
)

/**
 * The first-launch walkthrough.
 *
 * Shown once. Completion is recorded whether the user reads it or skips it, so
 * it never reappears on a later launch.
 */
@Composable
fun OnboardingScreen(onFinished: () -> Unit) {
    val container = LocalAppContainer.current
    val pagerState = rememberPagerState(pageCount = { onboardingPages.size })
    val scope = rememberCoroutineScope()

    fun finish() {
        container.appPreferences.onboardingComplete = true
        onFinished()
    }

    Box(Modifier.fillMaxSize()) {
        HorizontalPager(state = pagerState, modifier = Modifier.fillMaxSize()) { page ->
            OnboardingPageContent(onboardingPages[page])
        }

        Row(
            modifier = Modifier
                .align(Alignment.BottomCenter)
                .fillMaxWidth()
                .padding(horizontal = 24.dp, vertical = 36.dp),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            val isLast = pagerState.currentPage == onboardingPages.lastIndex

            TextButton(onClick = { finish() }, enabled = !isLast) {
                Text(if (isLast) "" else "Skip", color = AppColors.Muted)
            }

            Row(verticalAlignment = Alignment.CenterVertically) {
                repeat(onboardingPages.size) { index ->
                    val width by animateDpAsState(
                        targetValue = if (index == pagerState.currentPage) 18.dp else 8.dp,
                        animationSpec = tween(250),
                        label = "dotWidth",
                    )
                    Box(
                        modifier = Modifier
                            .padding(horizontal = 4.dp)
                            .size(width = width, height = 8.dp)
                            .background(
                                if (index == pagerState.currentPage) {
                                    AppColors.Accent
                                } else {
                                    Color.White.copy(alpha = 0.24f)
                                },
                                RoundedCornerShape(4.dp),
                            ),
                    )
                }
            }

            Button(
                onClick = {
                    if (isLast) {
                        finish()
                    } else {
                        scope.launch { pagerState.animateScrollToPage(pagerState.currentPage + 1) }
                    }
                },
                shape = RoundedCornerShape(20.dp),
                colors = ButtonDefaults.buttonColors(containerColor = AppColors.Accent),
            ) {
                Text(if (isLast) "Get started" else "Next", color = AppColors.Background)
            }
        }
    }
}

@Composable
private fun OnboardingPageContent(page: OnboardingPage) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(page.background)
            .padding(start = 32.dp, top = 80.dp, end = 32.dp, bottom = 100.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
    ) {
        Box(
            modifier = Modifier
                .size(120.dp)
                .background(Color.White.copy(alpha = 0.06f), CircleShape)
                .border(2.dp, AppColors.Accent, CircleShape),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                page.icon,
                contentDescription = null,
                tint = AppColors.Accent,
                modifier = Modifier.size(64.dp),
            )
        }
        Spacer(Modifier.height(40.dp))
        Text(
            page.title,
            style = MaterialTheme.typography.headlineMedium,
            fontWeight = FontWeight.Bold,
            textAlign = TextAlign.Center,
        )
        Spacer(Modifier.height(20.dp))
        Text(
            page.body,
            style = MaterialTheme.typography.bodyLarge,
            color = AppColors.Muted,
            textAlign = TextAlign.Center,
        )
    }
}
